### PR TITLE
Bring features/param-nullchecking up to date

### PR DESCRIPTION
**Pull request by kevinsun-dev**
In the process of updating `features/param-nullchecking` for updated specs (!!).

Changes:
 - Force merged master in (credit goes to @jaredpar)
 - Fixed several issues with param-nullchecking breaking async
 - Updated Tests

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 1]
```diff
@@ -254,7 +254,7 @@ public override BoundNode VisitLambda(BoundLambda node)
             try
             {
                 _factory.CurrentFunction = node.Symbol;
-                var visited = (BoundLambda)base.VisitLambda(node);
+                var visited = (BoundLambda)base.VisitLambda(node)!;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:257]: I would prefer a `Debug.Assert(visited is object);` instead of a suppression.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 2]
```diff
@@ -314,7 +314,7 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
                     _instrumenter = RemoveDynamicAnalysisInjectors(oldInstrumenter);
                 }
 
-                var visited = (BoundLocalFunctionStatement)base.VisitLocalFunctionStatement(node);
+                var visited = (BoundLocalFunctionStatement)base.VisitLocalFunctionStatement(node)!;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:317]: Same.

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs [hunk 3]
```diff
@@ -316,12 +316,9 @@ protected BoundStatement GenerateParameterStorage(LocalSymbol stateMachineVariab
                 }
             }
 
-            bodyBuilder.Add(GenerateStateMachineCreation(stateMachineVariable, frameType));
             var builtBody = bodyBuilder.ToImmutableAndFree();
             ImmutableArray<BoundStatement> newBody = LocalRewriter.ConstructNullCheckedStatementList(method.Parameters, builtBody, F);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs:320]: I know you didn't change this in this PR, but it would be good to rename `ConstructNullCheckedStatementList` to `TryConstructNullCheckedStatementList`, or something similar that conveys that the body it returns could be `default`, as it's not a nullable type.

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs [hunk 4]
```diff
@@ -1487,6 +1487,7 @@ internal ImmutableArray<BoundExpression> MakeTempsForDiscardArguments(ImmutableA
             return arguments;
         }
 
+#nullable disable
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:1490]: Too much churn for this PR? If so, please add a prototype to re-enable annotations, as this is all "new" code that we want to add to the codebase.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -10413,26 +10413,6 @@ private bool IsPossibleAnonymousMethodExpression()
                 tokenIndex++;
             }
 
-            if (precedence <= Precedence.Lambda)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10416]: Why is this moved? A mismerge or something?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10416]: Ah, looks like this is how it is in master.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs [hunk 6]
```diff
@@ -1206,13 +1208,13 @@ public static void Main()
...
     // Code size       29 (0x1d)
     .maxstack  3
-    IL_0000:  ldarg.1
-    IL_0001:  brtrue.s   IL_000e
-    IL_0003:  ldstr      ""s""
-    IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
-    IL_000d:  throw
-    IL_000e:  ldc.i4.s   -2
-    IL_0010:  newobj     ""C.<GetChars>d__0..ctor(int)""
+    IL_0000:  ldc.i4.s   -2
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs:1211]: This reordering seems incorrect. I wouldn't necessarily try to fix it in this PR, just prototype it. But we should be doing the null checks _first_ not after creating the enumerable closure object.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_INullCheckedParameters.cs [hunk 7]
```diff
@@ -12,7 +15,7 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
     public partial class IOperationTests : SemanticModelTestBase
     {
         [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
-        [Fact]
+        [Fact(Skip = "PROTOTYPE(BangBang)")]
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_INullCheckedParameters.cs:18]: Please update the baselines, don't skip them.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_INullCheckedParameters.cs [hunk 8]
```diff
@@ -88,37 +91,37 @@ public class C
...
+    Next (Regular) Block[B1]
+Block[B1] - Block
+    Predecessors: [B0]
+    Statements (0)
+    Jump if False (Regular) to Block[B3]
+        IBinaryOperation (BinaryOperatorKind.Equals) (OperationKind.Binary, Type: System.Boolean, IsImplicit) (Syntax: 'public void ... ing y!) { }')
+          Left: 
+            IParameterReferenceOperation: y (OperationKind.ParameterReference, Type: System.String, IsImplicit) (Syntax: 'public void ... ing y!) { }')
+          Right: 
+            ILiteralOperation (OperationKind.Literal, Type: System.String, Constant: null, IsImplicit) (Syntax: 'public void ... ing y!) { }')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_INullCheckedParameters.cs:105]: This constant value change also seems incorrect. Feel free to leave a prototype comment.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 9]
```diff
@@ -1482,7 +1482,7 @@ private void VisitNullChecks(IOperation operation, ImmutableArray<IParameterSymb
         private ConditionalOperation GenerateNullCheckForParameter(IParameterSymbol parameter, SyntaxNode syntax, SemanticModel semanticModel)
         {
             Debug.Assert(parameter.Language == LanguageNames.CSharp);
-            Optional<object> constantValue = default(Optional<object>);
+            ConstantValue constantValue = null;
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1485]: Actually, this is the source of the IOperation issue. It should be `ConstantValue.Null`: http://sourceroslyn.io/#Microsoft.CodeAnalysis/ConstantValue.cs,d8af96f1a97db377,references.

Comment (kevinsun-dev) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Thanks, updated as such.

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs (hunk 3 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs]: Renamed the method to better fit. Thanks.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs (hunk 1 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs]: Replaced the suppression with a `Debug.Assert`. An additional warning, but it should be fine. :+1:

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs (hunk 4 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:1490]: Added a PROTOTYPE comment 👍

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs (hunk 1 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs]: That additional warning seemed to become an error in testing:
`LocalRewriter.cs(257,31): error CS8600: Converting null literal or possible null value to non-nullable type.`
Reverted it back to what it was for now.

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:257]: It is a warning locally but we push all warnings to errors in the CI runs.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:257]: The warning is because you need to cast it to `BoundLambda?`, not `BoundLambda`.
